source/intel/intrin/crc32c_sse42_avx512.c

```diff
@@ -0,0 +1,194 @@
+    /*
+     * Definitions of the bit-reflected domain constants k1,k2,k3,k4,k5,k6
+     * are similar to those given at the end of the paper
+     *
+     * k1 = ( x ^ ( 512 * 4 + 32 ) mod P(x) << 32 )' << 1
+     * k2 = ( x ^ ( 512 * 4 - 32 ) mod P(x) << 32 )' << 1
+     * k3 = ( x ^ ( 512 + 32 ) mod P(x) << 32 )' << 1
+     * k4 = ( x ^ ( 512 - 32 ) mod P(x) << 32 )' << 1
+     * k5 = ( x ^ ( 128 + 32 ) mod P(x) << 32 )' << 1
+     * k6 = ( x ^ ( 128 - 32 ) mod P(x) << 32 )' << 1
+     */
+
+    static zalign_8 k1k2[8] = 
+        {0xdcb17aa4, 0xb9e02b86, 0xdcb17aa4, 0xb9e02b86, 0xdcb17aa4, 0xb9e02b86, 0xdcb17aa4, 0xb9e02b86};
+
+    static zalign_8 k3k4[8] =
+        {0x740eef02, 0x9e4addf8, 0x740eef02, 0x9e4addf8, 0x740eef02, 0x9e4addf8, 0x740eef02, 0x9e4addf8};
+    static zalign_2 k5k6[2] = {0xf20c0dfe, 0x14cd00bd6};
+    static zalign_2 k7k8[2] = {0xdd45aab8, 0x000000000};
+    static zalign_2 poly[2] = {0x105ec76f1, 0xdea713f1};
+
+    __m512i x0, x1, x2, x3, x4, x5, x6, x7, x8, y5, y6, y7, y8;
+    __m128i a0, a1, a2, a3;
+
+    /*
+     * There's at least one block of 256.
+     */
+    x1 = _mm512_loadu_si512((__m512i *)(input + 0x00));
+    x2 = _mm512_loadu_si512((__m512i *)(input + 0x40));
+    x3 = _mm512_loadu_si512((__m512i *)(input + 0x80));
+    x4 = _mm512_loadu_si512((__m512i *)(input + 0xC0));
+
+    x1 = _mm512_xor_si512(x1, _mm512_castsi128_si512(_mm_cvtsi32_si128(crc)));
+
+    x0 = _mm512_load_si512((__m512i *)k1k2);
+
+    input += 256;
+    length -= 256;
+
+    /*
+     * Parallel fold blocks of 256, if any.
+     */
+    while (length >= 256) {
+        x5 = _mm512_clmulepi64_epi128(x1, x0, 0x00);
+        x6 = _mm512_clmulepi64_epi128(x2, x0, 0x00);
+        x7 = _mm512_clmulepi64_epi128(x3, x0, 0x00);
+        x8 = _mm512_clmulepi64_epi128(x4, x0, 0x00);
+
+        x1 = _mm512_clmulepi64_epi128(x1, x0, 0x11);
+        x2 = _mm512_clmulepi64_epi128(x2, x0, 0x11);
+        x3 = _mm512_clmulepi64_epi128(x3, x0, 0x11);
+        x4 = _mm512_clmulepi64_epi128(x4, x0, 0x11);
+
+        y5 = _mm512_loadu_si512((__m512i *)(input + 0x00));
+        y6 = _mm512_loadu_si512((__m512i *)(input + 0x40));
+        y7 = _mm512_loadu_si512((__m512i *)(input + 0x80));
+        y8 = _mm512_loadu_si512((__m512i *)(input + 0xC0));
+
+        x1 = _mm512_xor_si512(x1, x5);
+        x2 = _mm512_xor_si512(x2, x6);
+        x3 = _mm512_xor_si512(x3, x7);
+        x4 = _mm512_xor_si512(x4, x8);
+
+        x1 = _mm512_xor_si512(x1, y5);
+        x2 = _mm512_xor_si512(x2, y6);
+        x3 = _mm512_xor_si512(x3, y7);
+        x4 = _mm512_xor_si512(x4, y8);
+
+        input += 256;
+        length -= 256;
+    }
+
+    /*
+     * Fold into 512-bits.
+     */
+    x0 = _mm512_load_si512((__m512i *)k3k4);
+
+    x5 = _mm512_clmulepi64_epi128(x1, x0, 0x00);
+    x1 = _mm512_clmulepi64_epi128(x1, x0, 0x11);
+    x1 = _mm512_xor_si512(x1, x2);
+    x1 = _mm512_xor_si512(x1, x5);
```

Comment:
The values that start out in x1 are effectively being multiplied three times by k3k4. (Similarly, x2 is multiplied twice, x3 once). Would we be able to reduce the dependency chain here by having three premultiplied constants instead of one, and XORing at the end?

Reply:
Note that having three premultiplied constants does increase cache usage - we could avoid this by multiplying k3k4 multiple times into each lane instead. That is:

```
x4 = MUL128(k3k4, k3k4) // CLMUL * 2, XOR; lat=5, 1/tp=2 (contends on port 5 for CLMUL)
x1 = MUL128(x1, k3k4)
x3 = MUL128(x3, k3k4)
x2 = MUL128(x2, x4) // waits 2 cycles for CLMUL latency
x1 = MUL128(x1, x4) // waits 2 cycles for CLMUL latency

x1 = (x2 ^ x3) ^ x1 // the first XOR can execute in parallel with the last MUL128
```

While this requires more multiplications (10 CLMULs vs 6), because CLMULs have a high latency, this approach might end up better utilizing the CLMUL capacity of the CPU.

Reply:
On further thought the premultiplication trick might not work, since we're not doing a real 128-bit multiplication operation. It's probably simpler to just load multiple constants.